L1Trigger/Phase2L1ParticleFlow/src/PFAlgo2HGC.cc

```diff
@@ -0,0 +1,648 @@
+            "requirement had been relaxed\n",
+            itk,
+            tk.floatPt(),
+            ibest,
+            r.calo[ibest].floatPt(),
+            drbest);
+    }
+  }
+}
+
+void PFAlgo2HGC::sum_tk2calo(Region &r,
+                             const std::vector<int> &tk2calo,
+                             std::vector<int> &calo2ntk,
+                             std::vector<float> &calo2sumtkpt,
+                             std::vector<float> &calo2sumtkpterr) const {
+  // for each calo, compute the sum of the track pt
+  for (int ic = 0, nc = r.calo.size(); ic < nc; ++ic) {
+    const auto &calo = r.calo[ic];
+    if (r.globalAbsEta(calo.floatEta()) > 2.5)
+      continue;
+    for (int itk = 0, ntk = r.track.size(); itk < ntk; ++itk) {
+      if (tk2calo[itk] == ic) {
+        const auto &tk = r.track[itk];
+        if (tk.muonLink || tk.used)
+          continue;
+        calo2ntk[ic]++;
+        calo2sumtkpt[ic] += tk.floatPt();
+        calo2sumtkpterr[ic] += std::pow(tk.floatCaloPtErr(), sumTkCaloErr2_ ? 2 : 1);
+      }
+    }
+    if (sumTkCaloErr2_ && calo2sumtkpterr[ic] > 0)
+      calo2sumtkpterr[ic] = std::sqrt(calo2sumtkpterr[ic]);
+  }
+}
+
+void PFAlgo2HGC::unlinkedtk_algo(Region &r, const std::vector<int> &tk2calo) const {
+  // in the meantime, promote unlinked low pt tracks to hadrons
+  for (int itk = 0, ntk = r.track.size(); itk < ntk; ++itk) {
+    auto &tk = r.track[itk];
+    if (tk2calo[itk] != -1 || tk.muonLink || tk.used)
+      continue;
+    float maxPt = (tk.hwStubs >= tightTrackMinStubs_ && tk.hwChi2 < 10. * tightTrackMaxChi2_)
+                      ? tightTrackMaxInvisiblePt_
+                      : maxInvisiblePt_;
+    if (tk.floatPt() < maxPt) {
+      if (debug_)
+        printf("PFAlgo2HGC \t track %3d (pt %7.2f) not matched to calo, kept as charged hadron\n", itk, tk.floatPt());
+      auto &p = addTrackToPF(r, tk);
+      p.hwStatus = GoodTK_NoCalo;
+      tk.used = true;
+    } else {
+      if (debug_)
+        printf("PFAlgo2HGC \t track %3d (pt %7.2f) not matched to calo, dropped\n", itk, tk.floatPt());
+      //discardTrack(r, tk, BadTK_NoCalo); // log this as discarded, for debugging
+    }
+  }
+}
+
+void PFAlgo2HGC::calo_relink(Region &r,
+                             const std::vector<int> &calo2ntk,
+                             const std::vector<float> &calo2sumtkpt,
+                             const std::vector<float> &calo2sumtkpterr) const {
+  /// OPTIONAL STEP: try to recover split hadron showers (v1.0):
+  //     take hadrons that are not track matched, close by a hadron which has an excess of track pt vs calo pt
+  //     add this pt to the calo pt of the other cluster
+  //     off by default, as it seems to not do much in jets even if it helps remove tails in single-pion events
+  std::vector<float> addtopt(r.calo.size(), 0);
+  for (int ic = 0, nc = r.calo.size(); ic < nc; ++ic) {
+    auto &calo = r.calo[ic];
```

Comment:
const auto&

Reply:
it can't be const, it may be modified later in the code (there's a `calo.used = true` near the end of the block)